Common/NearestNeighbour/src/main/java/org/tribuo/common/nearest/KNNTrainer.java

```diff
@@ -114,18 +120,22 @@ private KNNTrainer() {}
      * @param numThreads The number of threads to use.
      * @param combiner The combination function to aggregate the k predictions.
      * @param backend The computational backend.
+     * @param nqFactoryType The nearest neighbour implementation factory to use.
      */
-    public KNNTrainer(int k, DistanceType distType, int numThreads, EnsembleCombiner<T> combiner, Backend backend) {
+    public KNNTrainer(int k, DistanceType distType, int numThreads, EnsembleCombiner<T> combiner,
+                      Backend backend, NeighboursQueryFactoryType nqFactoryType) {
         this.k = k;
         this.distType = distType;
         this.numThreads = numThreads;
         this.combiner = combiner;
         this.backend = backend;
+        this.neighboursQueryFactory = NeighboursQueryFactoryType.getNeighboursQueryFactory(nqFactoryType, distType, numThreads);
```

Comment:
Same comment as HDBSCAN, there should be a constructor that takes a factory directly so it's as flexible as the configuration system.